shared/dataflow/codeql/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -958,6 +961,377 @@ module MakeImplCommon<InputSig Lang> {
       TApproxAccessPathFrontSome(ApproxAccessPathFront apf)
   }
 
+  bindingset[t1, t2]
+  pragma[inline_late]
+  private predicate typeStrongerThan0(DataFlowType t1, DataFlowType t2) { typeStrongerThan(t1, t2) }
+
+  private predicate callEdge(DataFlowCall call, DataFlowCallable c, ArgNode arg, ParamNode p) {
+    viableParamArg(call, p, arg) and
+    c = getNodeEnclosingCallable(p)
+  }
+
+  signature module TypeFlowInput {
+    predicate enableTypeFlow();
+
+    /** Holds if the edge is possibly needed in the direction `call` to `c`. */
+    predicate relevantCallEdgeIn(DataFlowCall call, DataFlowCallable c);
+
+    /** Holds if the edge is possibly needed in the direction `c` to `call`. */
+    predicate relevantCallEdgeOut(DataFlowCall call, DataFlowCallable c);
+
+    /**
+     * Holds if the edge is followed in data flow in the direction `call` to `c`
+     * and the call context `cc`.
+     */
+    predicate dataFlowTakenCallEdgeIn(DataFlowCall call, DataFlowCallable c, boolean cc);
```

Comment:
I don't think it matters _that_ much. But it does correlate very strongly with whether or not we have interesting type bounds, so it ought to be almost for free and does increase precision a bit. I.e. if `cc` is false then we generally know nothing and we'll just have a single type-tuple per tracked node, so if we conflate those with the `cc=true` case then we just ruin all of the type bounds we'd otherwise have in that case.

The dual question about whether it would make sense to increase the call context precision here is a big maybe, but we can say for sure that it brings a significant additional risk, so I don't we should. My gut feeling is that it wouldn't add that much precision anyway, since this analysis exactly brings some of the long-range type restrictions that the 1-level call context that we otherwise employ doesn't yield. And extending the call context a little bit to improve the long-range type restrictions seems like it won't do much.